File: CS/E724/CS/ViewHotTrackHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace HotTrack {
    using System;
    using DevExpress.XtraGrid;
    using DevExpress.XtraGrid.Columns;
    using DevExpress.XtraGrid.Views.Grid;
    using DevExpress.XtraGrid.Views.Grid.ViewInfo;
    using System.Drawing;

    namespace dxSample {
        public enum GridHotTrackMode { Cell, Row };
        public class GridHotTrackHelper : IDisposable {
            GridHotTrackMode mode;
            GridView targetView;
            Color hotTrackColor;
            public GridHotTrackHelper(GridView targetView, Color hotTrackColor, GridHotTrackMode mode = GridHotTrackMode.Cell) {
                this.mode = mode;
                this.targetView = targetView;
                this.hotTrackColor = hotTrackColor;
                Subscribe();
            }
            public Color HotTrackColor {
                get { return hotTrackColor; }
                set { hotTrackColor = value; }
            }
            protected void Subscribe() {
                targetView.MouseMove += TargetView_MouseMove;
                targetView.MouseLeave += TargetView_MouseLeave;
                targetView.ShownEditor += TargetView_ShownEditor;
                targetView.RowCellStyle += TargetView_RowCellStyle;
            }
            protected void Unsubscribe() {
                if (targetView == null) return;
                SetHotTrack(GridControl.InvalidRowHandle);
                targetView.MouseMove -= TargetView_MouseMove;
                targetView.MouseLeave -= TargetView_MouseLeave;
                targetView.ShownEditor -= TargetView_ShownEditor;
                targetView.RowCellStyle -= TargetView_RowCellStyle;
            }
            int hotTrackedRow = GridControl.InvalidRowHandle;
            GridColumn hotTrackedColumn = null;
            void SetHotTrack(int rowHandle, GridColumn column = null) {
                if (hotTrackedRow != rowHandle || hotTrackedColumn != column) {
                    RefreshAppearance(hotTrackedRow, hotTrackedColumn);
                }
                this.hotTrackedRow = rowHandle;
                this.hotTrackedColumn = column;
                RefreshAppearance(hotTrackedRow, hotTrackedColumn);
            }
            void RefreshAppearance(int rowHandle, GridColumn column) {
                if (rowHandle == GridControl.InvalidRowHandle) return;
                if (targetView == null) return;
                var viewInfo = targetView.GetViewInfo() as GridViewInfo;
                if (viewInfo == null) return;
                var rowInfo = viewInfo.GetGridRowInfo(rowHandle);
                if (rowInfo != null) RefreshAppearance(viewInfo, rowInfo, column);
                targetView.InvalidateRow(rowHandle);
            }
            void RefreshAppearance(GridViewInfo viewInfo, GridRowInfo rowInfo, GridColumn column) {
                if (column != null) {
                    GridDataRowInfo dataRow = rowInfo as GridDataRowInfo;
                    if (dataRow != null) {
                        var cell = dataRow.Cells[column];
                        if (cell != null) cell.State = DevExpress.XtraGrid.Views.Base.GridRowCellState.Dirty;
                    }
                }
                else {
                    rowInfo.SetDataDirty();
                }
            }
            void TargetView_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e) {
                if (targetView.IsEditing) return;
                var hitInfo = targetView.CalcHitInfo(e.X, e.Y);
                SetHotTrack(hitInfo.RowHandle, mode == GridHotTrackMode.Cell ? hitInfo.Column : null);
            }
            void TargetView_MouseLeave(object sender, EventArgs e) { SetHotTrack(GridControl.InvalidRowHandle); }
            void TargetView_ShownEditor(object sender, EventArgs e) { SetHotTrack(GridControl.InvalidRowHandle); }
            void TargetView_RowCellStyle(object sender, RowCellStyleEventArgs e) {
                if (e.RowHandle == hotTrackedRow) {
                    if (mode == GridHotTrackMode.Cell) {
                        if (hotTrackedColumn == e.Column && e.Column != null) e.Appearance.BackColor = HotTrackColor;
                    }
                    else
                        e.Appearance.BackColor = HotTrackColor;
                }
            }
            public void Dispose() {
                Unsubscribe();
                targetView = null;
            }
        }
    }
}

File: VB/E724/CS/ViewHotTrackHelper.vb
Imports System.Drawing
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace HotTrack

    Namespace dxSample
        Public Enum GridHotTrackMode
            Cell
            Row
        End Enum
        Public Class GridHotTrackHelper
            Implements IDisposable

            Private mode As GridHotTrackMode
            Private targetView As GridView

            Private hotTrackColor_Renamed As Color
            Public Sub New(ByVal targetView As GridView, ByVal hotTrackColor As Color, Optional ByVal mode As GridHotTrackMode = GridHotTrackMode.Cell)
                Me.mode = mode
                Me.targetView = targetView
                Me.hotTrackColor_Renamed = hotTrackColor
                Subscribe()
            End Sub
            Public Property HotTrackColor() As Color
                Get
                    Return hotTrackColor_Renamed
                End Get
                Set(ByVal value As Color)
                    hotTrackColor_Renamed = value
                End Set
            End Property
            Protected Sub Subscribe()
                AddHandler targetView.MouseMove, AddressOf TargetView_MouseMove
                AddHandler targetView.MouseLeave, AddressOf TargetView_MouseLeave
                AddHandler targetView.ShownEditor, AddressOf TargetView_ShownEditor
                AddHandler targetView.RowCellStyle, AddressOf TargetView_RowCellStyle
            End Sub
            Protected Sub Unsubscribe()
                If targetView Is Nothing Then
                    Return
                End If
                SetHotTrack(GridControl.InvalidRowHandle)
                RemoveHandler targetView.MouseMove, AddressOf TargetView_MouseMove
                RemoveHandler targetView.MouseLeave, AddressOf TargetView_MouseLeave
                RemoveHandler targetView.ShownEditor, AddressOf TargetView_ShownEditor
                RemoveHandler targetView.RowCellStyle, AddressOf TargetView_RowCellStyle
            End Sub
            Private hotTrackedRow As Integer = GridControl.InvalidRowHandle
            Private hotTrackedColumn As GridColumn = Nothing
            Private Sub SetHotTrack(ByVal rowHandle As Integer, Optional ByVal column As GridColumn = Nothing)
                If hotTrackedRow <> rowHandle OrElse hotTrackedColumn IsNot column Then
                    RefreshAppearance(hotTrackedRow, hotTrackedColumn)
                End If
                Me.hotTrackedRow = rowHandle
                Me.hotTrackedColumn = column
                RefreshAppearance(hotTrackedRow, hotTrackedColumn)
            End Sub
            Private Sub RefreshAppearance(ByVal rowHandle As Integer, ByVal column As GridColumn)
                If rowHandle = GridControl.InvalidRowHandle Then
                    Return
                End If
                If targetView Is Nothing Then
                    Return
                End If
                Dim viewInfo = TryCast(targetView.GetViewInfo(), GridViewInfo)
                If viewInfo Is Nothing Then
                    Return
                End If
                Dim rowInfo = viewInfo.GetGridRowInfo(rowHandle)
                If rowInfo IsNot Nothing Then
                    RefreshAppearance(viewInfo, rowInfo, column)
                End If
                targetView.InvalidateRow(rowHandle)
            End Sub
            Private Sub RefreshAppearance(ByVal viewInfo As GridViewInfo, ByVal rowInfo As GridRowInfo, ByVal column As GridColumn)
                If column IsNot Nothing Then
                    Dim dataRow As GridDataRowInfo = TryCast(rowInfo, GridDataRowInfo)
                    If dataRow IsNot Nothing Then
                        Dim cell = dataRow.Cells(column)
                        If cell IsNot Nothing Then
                            cell.State = DevExpress.XtraGrid.Views.Base.GridRowCellState.Dirty
                        End If
                    End If
                Else
                    rowInfo.SetDataDirty()
                End If
            End Sub
            Private Sub TargetView_MouseMove(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
                If targetView.IsEditing Then
                    Return
                End If
                Dim hitInfo = targetView.CalcHitInfo(e.X, e.Y)
                SetHotTrack(hitInfo.RowHandle,If(mode = GridHotTrackMode.Cell, hitInfo.Column, Nothing))
            End Sub
            Private Sub TargetView_MouseLeave(ByVal sender As Object, ByVal e As EventArgs)
                SetHotTrack(GridControl.InvalidRowHandle)
            End Sub
            Private Sub TargetView_ShownEditor(ByVal sender As Object, ByVal e As EventArgs)
                SetHotTrack(GridControl.InvalidRowHandle)
            End Sub
            Private Sub TargetView_RowCellStyle(ByVal sender As Object, ByVal e As RowCellStyleEventArgs)
                If e.RowHandle = hotTrackedRow Then
                    If mode = GridHotTrackMode.Cell Then
                        If hotTrackedColumn Is e.Column AndAlso e.Column IsNot Nothing Then
                            e.Appearance.BackColor = HotTrackColor
                        End If
                    Else
                        e.Appearance.BackColor = HotTrackColor
                    End If
                End If
            End Sub
            Public Sub Dispose() Implements IDisposable.Dispose
                Unsubscribe()
                targetView = Nothing
            End Sub
        End Class
    End Namespace
End Namespace
